File: plugins/set_value/tests/Python/set_value_test_04.py
#!/usr/bin/env python3

# SPDX-FileCopyrightText: 2020 Institute of Bioorganic Chemistry Polish Academy of Science (PSNC)
#
# SPDX-License-Identifier: BSD-3-Clause

from sys import stdout, argv, exit
import pdi, yaml
import numpy as np

config_path = argv[1]

with open(config_path, 'r') as config_file:
	try:    
		config = yaml.safe_load(config_file)
	except yaml.YAMLError as exc:
		exit(exc)

pdi.init(yaml.dump(config))

# read values
outer_array = pdi.access("outer_array", pdi.IN)
for i in range(4):
	print("outer_array[" + str(i) + "].scalar == " + str(outer_array[i].scalar))
	assert outer_array[i].scalar == i * 10
	for j in range(4):
		print("outer_array[" + str(i) + "].inner_array[" + str(j) + "] == " + str(outer_array[i].inner_array[j]))
		assert outer_array[i].inner_array[j] ==  i * 10 + j + 1
pdi.release("outer_array")

# write values
outer_array = pdi.access("outer_array", pdi.OUT)
for i in range(4):
	print("outer_array[" + str(i) + "].scalar = " + str(i * 100))
	outer_array[i].scalar = i * 100
	for j in range(4):
		print("outer_array[" + str(i) + "].inner_array[" + str(j) + "] = " + str(i * 100 + j + 10))
		outer_array[i].inner_array[j] =  i * 100 + j + 10
pdi.release("outer_array")

# read new values
outer_array = pdi.access("outer_array", pdi.IN)
for i in range(4):
	print("outer_array[" + str(i) + "].scalar == " + str(outer_array[i].scalar))
	assert outer_array[i].scalar == i * 100
	for j in range(4):
		print("outer_array[" + str(i) + "].inner_array[" + str(j) + "] == " + str(outer_array[i].inner_array[j]))
		assert outer_array[i].inner_array[j] ==  i * 100 + j + 10
pdi.release("outer_array")

File: plugins/set_value/tests/set_value_test_04.c
/*
 * SPDX-FileCopyrightText: 2020 Institute of Bioorganic Chemistry Polish Academy of Science (PSNC)
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

#include <assert.h>

#include <paraconf.h>
#include <pdi.h>

int main(int argc, char* argv[])
{
	PDI_init(PC_parse_path(argv[1]));

	// int
	int* size;
	PDI_access("array_size", (void**)&size, PDI_IN);
	assert(*size == 6);
	PDI_release("array_size");

	int* int_array;
	PDI_access("array_metadata", (void**)&int_array, PDI_IN);
	for (int i = 0; i < 4; i++) {
		assert(int_array[i] == i + 1);
	}
	PDI_release("array_metadata");

	PDI_event("update_array");

	PDI_access("array_metadata", (void**)&int_array, PDI_IN);
	for (int i = 0; i < 6; i++) {
		assert(int_array[i] == 9 - i);
	}
	PDI_release("array_metadata");

	PDI_access("array_size", (void**)&size, PDI_IN);
	assert(*size == 0);
	PDI_reclaim("array_size");

	free(size);

	PDI_finalize();
}
